oshmem/mca/scoll/basic/scoll_basic_alltoall.c

```diff
@@ -0,0 +1,112 @@
+ *                         All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "oshmem_config.h"
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "oshmem/constants.h"
+#include "oshmem/op/op.h"
+#include "oshmem/mca/spml/spml.h"
+#include "oshmem/mca/scoll/scoll.h"
+#include "oshmem/mca/scoll/base/base.h"
+#include "scoll_basic.h"
+
+static int _algorithm_simple(struct oshmem_group_t *group,
+                             void *target,
+                             const void *source,
+                             ptrdiff_t dst, ptrdiff_t sst,
+                             size_t nelems,
+                             size_t element_size,
+                             long *pSync);
+
+int mca_scoll_basic_alltoall(struct oshmem_group_t *group,
+                             void *target,
+                             const void *source,
+                             ptrdiff_t dst, ptrdiff_t sst,
+                             size_t nelems,
+                             size_t element_size,
+                             long *pSync,
+                             int alg)
+{
+    int rc = OSHMEM_SUCCESS;
+
+    /* Arguments validation */
+    if (!group) {
+        SCOLL_ERROR("Active set (group) of PE is not defined");
+        rc = OSHMEM_ERR_BAD_PARAM;
+    }
+
+    /* Check if this PE is part of the group */
+    if ((rc == OSHMEM_SUCCESS) && oshmem_proc_group_is_member(group)) {
+        int i = 0;
+
+        if (pSync) {
+            rc = _algorithm_simple(group,
+                                   target,
+                                   source,
+                                   dst,
+                                   sst,
+                                   nelems,
+                                   element_size,
+                                   pSync);
+        } else {
+            SCOLL_ERROR("Incorrect argument pSync");
+            rc = OSHMEM_ERR_BAD_PARAM;
+        }
+
+        /* Restore initial values */
+        SCOLL_VERBOSE(12,
+                      "PE#%d Restore special synchronization array",
+                      group->my_pe);
+        for (i = 0; pSync && (i < _SHMEM_ALLTOALL_SYNC_SIZE); i++) {
+            pSync[i] = _SHMEM_SYNC_VALUE;
+        }
+    }
+
+    return rc;
+}
+
+static int _algorithm_simple(struct oshmem_group_t *group,
+                             void *target,
+                             const void *source,
+                             ptrdiff_t tst, ptrdiff_t sst,
+                             size_t nelems,
+                             size_t element_size,
+                             long *pSync)
+{
+    int rc = OSHMEM_SUCCESS;
+    int pe_cur;
+    int i;
+    int j;
+
+    SCOLL_VERBOSE(14,
+                  "[#%d] send data to all PE in the group",
+                  group->my_pe);
+    j = oshmem_proc_group_find_id(group, group->my_pe);
+    for (i = 0; (i < group->proc_count) && (rc == OSHMEM_SUCCESS); i++) {
```

Comment:
allso, maybe for better distribution of traffic, every PE should start from itself and not from 0, to avoid many to one.
see example here;
https://github.com/deffjammer/mmu_notifier_callouts/blob/master/traffic-apps/mpp_bench_v1.0_sgi/mpp_bench.c#L286

@yosefe - please review it as well